compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEUnrollAndDistributeWorkgroup.cpp

```diff
@@ -0,0 +1,503 @@
+      if (!op) return operand.get() == forOp.getInductionVar();
+
+      // Check fo induction var and whether the parent scf.for is the same as
+      // the one we're using for the hoist
+      auto parentForOp = op->getParentOfType<scf::ForOp>();
+      return (operand.get() == forOp.getInductionVar()) ||
+             (parentForOp && parentForOp == forOp);
+    };
+
+    // Collect the internal logical objectfifos based on dma ops' sources and
+    // targets. Then walk all dma ops and try to hoist them either before or
+    // after the scf.for depending on whether target or source logical
+    // objectfifo is an internal one.
+    DenseSet<AMDAIE::LogicalObjectFifoFromMemrefOp> internalLogicalObjectFifos;
+    collectInternalLogicalObjectFifos(forOp, internalLogicalObjectFifos);
+    forOp.walk([&](AMDAIE::DmaCpyNdOp dmaOp) {
+      if ((!internalLogicalObjectFifos.contains(dmaOp.getSourceObjectFifo()) ||
+           !internalLogicalObjectFifos.contains(dmaOp.getTargetObjectFifo())) &&
+          !llvm::any_of(dmaOp->getOpOperands(), [&](OpOperand &operand) {
+            return dependsOnLoop(operand);
+          })) {
+        if (internalLogicalObjectFifos.contains(dmaOp.getTargetObjectFifo()))
+          rewriter.moveOpBefore(dmaOp, forOp);
+        if (internalLogicalObjectFifos.contains(dmaOp.getSourceObjectFifo()))
+          rewriter.moveOpAfter(dmaOp, forOp);
+        hoistHappened = true;
+      }
+      return WalkResult::advance();
+    });
+    if (hoistHappened) return success();
+    return failure();
+  }
+
+  /// Unroll the for loop and while doing so, create new clones of
+  /// AMDAIE::LogicalObjectFifoFromMemrefOp so they can be distributed
+  /// onto multiple physical locations later.
+  LogicalResult loopUnrollAndDistributeLogicalObjectFifos(
+      RewriterBase &rewriter, scf::ForOp forOp) const {
+    Block *loopBodyBlock = forOp.getBody();
+    auto builder = OpBuilder::atBlockTerminator(loopBodyBlock);
+
+    // Keep a pointer to the last non-terminator operation in the original block
+    // so that we know what to clone (since we are doing this in-place).
+    Block::iterator srcBlockEnd = std::prev(loopBodyBlock->end(), 2);
+
+    // Update loop bounds
+    auto lbInt = getConstantIntValue(forOp.getLowerBound()).value();
+    auto ubInt = getConstantIntValue(forOp.getUpperBound()).value();
+    auto stepInt = getConstantIntValue(forOp.getStep()).value();
+    if (lbInt != 0 || stepInt != 1) return failure();
+    if (stepInt == (ubInt - lbInt)) return failure();
+    Value forOpIV = forOp.getInductionVar();
+    forOp.setLowerBound(
+        rewriter.create<arith::ConstantIndexOp>(rewriter.getUnknownLoc(), 0));
+    forOp.setUpperBound(
+        rewriter.create<arith::ConstantIndexOp>(rewriter.getUnknownLoc(), 1));
+    forOp.setStep(
+        rewriter.create<arith::ConstantIndexOp>(rewriter.getUnknownLoc(), 1));
+
+    // Iterate through the loop and create body
```

Comment:
Maybe add here that we're skipping the first iteration of the loop ?

Reply:
Added it.